cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/Dag.java

```diff
@@ -168,11 +168,19 @@ void validate() {
     Set<String> childStopperNodes = Sets.union(sinks, conditionNodes);
     for (String condition : conditionNodes) {
       Set<String> parentNodes = parentsOf(condition, parentStopperNodes);
-      Set<String> removeConditions = Sets.difference(parentNodes, conditionNodes);
-      // If the parentNodes only contain conditions, then do not add create subdag for them.
-      // This particular subdag will be created when we go from parent condition to child in the for loop below.
-      if (!removeConditions.isEmpty()) {
-        dags.add(createSubDag(accessibleFrom(removeConditions, childStopperNodes)));
+      Set<String> parentsWithoutConditions = Sets.difference(parentNodes, conditionNodes);
+      // If the parentNodes only contain conditions, then do not add create subdag for them, because subdag for them
+      // will be will be created when we go from parent condition to child in the for loop below.
+      if (!parentsWithoutConditions.isEmpty()) {
+        // Parent nodes not only contain conditions but other nodes as well
+        // Add back those conditions as well while creating subdag
+        // For example assume following dag where c1 and c4 are conditions/
+        // n1-n2-c1-n3-n4-c4-n5
+        // Now for condition node c4 we get parent nodes with stop nodes as conditions and sources - <c1,n3,n4>
+        // Now we get parentsWithoutConditions as <n3, n4> to make sure parent nodes not contains only condition nodes
+        // However to create subdag we still need c1 and c4 nodes. c4 gets added as a part of accessibleFrom method
+        // and c1 gets added back as a part of union.
+        dags.add(createSubDag(Sets.union(accessibleFrom(parentsWithoutConditions, childStopperNodes), parentNodes)));
```

Comment:
Dont think its a bug. Its an intended behavior where we create 4 subdags since we are splitting based on the condition, which is handled while generating phases. I have added test case for this scenario. Please let me know if any issues.